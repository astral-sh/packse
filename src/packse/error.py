import textwrap
from pathlib import Path


class PackseError(Exception):
    """Base type for all Packse exceptions"""


class UserError(PackseError):
    """It's the user's fault :)"""


class RequiresExtra(UserError):
    def __init__(self, feature: str, extra: str) -> None:
        message = f"Using {feature} requires installation with extra {extra!r}"
        super().__init__(message)


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"File at '{path}' is not a valid scenario: {reason}"
        super().__init__(message)


class InvalidPackageVersion(UserError):
    """Version is not PEP compliant"""

    def __init__(self, version: str) -> None:
        message = f"Version {version!r} is not valid."
        super().__init__(message)


class FileNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishError(PackseError):
    pass


class PublishToolError(PublishError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishNoCredentials(PublishError):
    def __init__(self) -> None:
        message = (
            "No credentials found for publish!\n"
            "Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`"
        )
        super().__init__(message)


class PublishAlreadyExists(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish for '{package}' already exists"
        super().__init__(message)


class PublishRateLimit(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to rate limits"
        super().__init__(message)


class PublishConnectionError(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to a connection error"
        super().__init__(message)


class InvalidPublishTarget(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"Publish target at '{path}' is not a valid: {reason}"
        super().__init__(message)


class ServeError(PackseError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeThreadError(ServeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeCommandError(ServeError):
    def __init__(self, message: str, stderr: str) -> None:
        if stderr.strip():
            message += f":\n\n{stderr}"
        super().__init__(message)


class ServeAddressInUse(ServeError):
    def __init__(self, url: str) -> None:
        message = f"Address '{url}' already in use"
        super().__init__(message)


class ServeAlreadyRunning(ServeError):
    def __init__(self) -> None:
        message = "Server is already running"
        super().__init__(message)
