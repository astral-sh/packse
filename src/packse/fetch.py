import logging
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

import pkg_resources

from packse.error import DestinationAlreadyExists

logger = logging.getLogger(__name__)


def fetch(
    dest: Path | None = None,
    ref: str | None = None,
    repo_url: str = "https://github.com/astral-sh/packse",
    repo_subdir: str = "scenarios",
    force: bool = False,
):
    start_time = time.time()
    dest = dest or (Path.cwd() / "scenarios")

    if dest.exists() and not force:
        raise DestinationAlreadyExists(dest)

    if not ref:
        ref = pkg_resources.get_distribution("packse").version
        if ref == "0.0.0":
            ref = "HEAD"

    with tempfile.TemporaryDirectory() as tmpdir:
        # Perform a sparse checkout where we only grab the `scenarios` folder
        logger.info("Cloning repository %s", repo_url)
        subprocess.check_call(
            [
                "git",
                "clone",
                "-n",
                "--filter=tree:0",
                repo_url,
                "repo",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=tmpdir,
        )

        logger.info("Checking out directory '%s' at ref %s", repo_subdir, ref)
        subprocess.check_call(
            ["git", "sparse-checkout", "set", "--no-cone", repo_subdir],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "fetch", "origin", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "checkout", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )

        src = Path(tmpdir) / "repo" / repo_subdir
        file_count = 0
        for file in sorted(src.iterdir()):
            file_count += 1
            logger.info("Found %s", file.name)

        logger.info("Copying files into '%s'", dest)
        shutil.copytree(src, dest, dirs_exist_ok=True)

    logger.info(
        "Fetched %s files in %.2fs",
        file_count,
        time.time() - start_time,
    )
