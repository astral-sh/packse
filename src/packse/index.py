import errno
import hashlib
import importlib
import logging
import os
import shutil
import signal
import subprocess
import sys
import tempfile
import time
from contextlib import contextmanager, nullcontext
from pathlib import Path
from typing import Generator

from packse import __development_base_path__
from packse.build import build
from packse.error import (
    RequiresExtra,
    ServeAddressInUse,
    ServeAlreadyRunning,
    ServeCommandError,
)
from packse.inspect import inspect
from packse.template import create_from_template

logger = logging.getLogger(__name__)


def index_up(
    host: str = "localhost",
    port: int = 3141,
    background: bool = False,
    dist_dir: Path | None = None,
    reset: bool = False,
    offline: bool = False,
):
    try:
        importlib.import_module("pypiserver")
    except ImportError:
        raise RequiresExtra("index commands", "index")

    if not dist_dir:
        dist_dir = Path.cwd() / "dist"
    else:
        dist_dir = Path(dist_dir).resolve()

    if background:
        if pid := get_server_pid():
            if is_running(pid):
                raise ServeAlreadyRunning()

    if reset:
        logger.info("Clearing existing server data...")
        if dist_dir.exists():
            shutil.rmtree(dist_dir)

    state_path = get_packse_state_path()

    logger.debug("Storing server metadata at %s", state_path)
    server_log_path = state_path / "server.log" if background else None
    index_url = f"http://{host}:{port}"
    index_note = "offline index" if offline else "local index with PyPI fallback"
    logger.info("Starting %s at %s", index_note, index_url)
    with start_index_server(
        dist_dir,
        host,
        port,
        server_log_path,
        offline=offline,
    ) as server_process:
        # Copy build dependencies over
        add_vendored_build_deps(dist_dir, offline)

        if background:
            logger.info("Running in background with pid %s", server_process.pid)
            logger.info("Stop index server with `packse index down`.")
            print(index_url)

        else:
            debug = logger.getEffectiveLevel() <= logging.DEBUG
            if not debug:
                logger.info("Hiding server output, use `-v` to stream server logs")

            logger.info("Ready! [Stop with Ctrl-C]")

            line = ""
            while True:
                line = server_process.stdout.readline().decode()
                if debug:
                    print(line, end="")


def index_down() -> bool:
    state_path = get_packse_state_path(create=False)

    if not state_path.exists():
        logger.warning("No server detected!")
        return False

    if pid := get_server_pid():
        if not is_running(pid):
            logger.info("Server looks shutdown already.")
            reset_pidfile()
            return

        logger.info("Stopping server with pid %s...", pid)
        try:
            os.kill(pid, signal.SIGINT)
        except ProcessLookupError:
            # Just move on if it's already dead
            pass

        # Wait for the server to exit for 10s
        limit = 10
        start_time = time.time()
        while time.time() - start_time < limit:
            if not is_running(pid):
                break
            time.sleep(0.1)
        else:
            # After the limit is reached, send a stronger signal...
            logger.info("Server has not stopped after %ds, sending kill...", limit)
            os.kill(pid, signal.SIGKILL)

            while is_running(pid):
                time.sleep(0.1)

        reset_pidfile()
        logger.info("Stopped server!")
        return True
    else:
        logger.warning("No server detected!")
        return False


@contextmanager
def start_index_server(
    dist_dir: Path,
    host: str,
    port: int,
    server_log_path: Path | None,
    offline: bool,
) -> Generator[subprocess.Popen, None, None]:
    server_url = f"http://{host}:{port}"
    dist_dir.mkdir(parents=True, exist_ok=True)

    command = [
        sys.executable,
        "-m",
        "pypiserver",
        "run",
        "--host",
        host,
        "--port",
        str(port),
        "-v",
        # Disable auth
        "-P",
        ".",
        "-a",
        ".",
        str(dist_dir),
    ]

    if offline:
        command.append("--disable-fallback")

    logger.debug("Running: %s", " ".join(command))

    server_process = subprocess.Popen(
        command,
        stderr=subprocess.STDOUT,
        stdout=server_log_path.open("wb") if server_log_path else subprocess.PIPE,
    )

    try:
        # Wait for server to be ready
        with server_log_path.open("rb") if server_log_path else nullcontext(
            server_process.stdout
        ) as server_output:
            line = ""
            lines = []

            # Watch for startup
            # TODO: Use a healthceck instead?
            while "Serving on" not in line and "Listening on" not in line:
                try:
                    server_process.wait(0)
                except subprocess.TimeoutExpired:
                    pass
                else:
                    # The process stopped already
                    if (
                        logger.getEffectiveLevel() > logging.DEBUG
                        and server_output.seekable()
                    ):
                        server_output.seek(0)
                    err_output = (
                        "\n".join(lines) + server_output.read().decode().strip()
                    )
                    raise ServeCommandError("Failed to start server!", err_output)

                line = server_output.readline().decode()
                if logger.getEffectiveLevel() <= logging.DEBUG:
                    print(line, end="")
                elif not server_output.seekable():
                    # Track the lines if we'll never be able to get them back
                    lines.append(line)
                if "Address already in use" in line:
                    raise ServeAddressInUse(server_url)

        # Server started!
        write_server_pid(server_process.pid)

        if server_log_path:
            logger.info("Writing server logs to %s", server_log_path)

        try:
            display_path = f"./{dist_dir.relative_to(Path.cwd())}"
        except Exception:
            display_path = dist_dir

        logger.info(f"Serving distributions from {display_path}")

        yield server_process

    except BaseException as exc:
        server_process.terminate()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        server_process.kill()

        stdout, _ = server_process.communicate(timeout=1)
        if logger.getEffectiveLevel() <= logging.DEBUG:
            print(stdout.decode())

        # Do not reset the pidfile when a server was already running!
        if not isinstance(exc, ServeAddressInUse):
            reset_pidfile()

        raise


def add_vendored_build_deps(dist_dir: Path, offline: bool):
    build_directory = __development_base_path__ / "vendor" / "build"
    if build_directory.exists():
        logger.debug("Copying vendored build dependencies...")
        shutil.copytree(build_directory, dist_dir / "build", dirs_exist_ok=True)
    else:
        logger.warning("No vendored build dependencies found at %s" % build_directory)
        if offline:
            logger.error(
                "Server has no fallback PyPI access and source distribution builds are likely to fail"
            )


def get_packse_state_path(create: bool = True) -> Path:
    path = Path(os.environ.get("PACKSE_STATE_PATH") or (Path.home() / ".packse"))
    if create:
        path.mkdir(exist_ok=True)
    return path


def get_pidfile_path() -> Path:
    return get_packse_state_path() / "server.pid"


def get_server_pid() -> int | None:
    pidfile = get_pidfile_path()
    if not pidfile.exists():
        return None
    else:
        logger.debug("Found existing pid at %s", pidfile)
        return int(pidfile.read_text())


def write_server_pid(pid: int) -> None:
    pidfile = get_pidfile_path()
    pidfile.write_text(str(pid))


def reset_pidfile():
    pidfile = get_pidfile_path()
    pidfile.unlink(missing_ok=True)


def is_running(pid):
    try:
        os.kill(pid, 0)
    except OSError as err:
        # ESRCH: pid not exist
        if err.errno == errno.ESRCH:
            return False
    return True


def sha256_file(path: Path):
    h = hashlib.sha256()

    with open(path, "rb") as file:
        while True:
            # Reading is buffered, so we can read smaller chunks.
            chunk = file.read(h.block_size)
            if not chunk:
                break
            h.update(chunk)

    return h.hexdigest()


def build_index(
    targets: list[Path],
    no_hash: bool,
    short_names: bool,
    dist_dir: Path | None,
):
    out_path = Path("./index")
    if out_path.exists():
        shutil.rmtree(out_path)

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir = Path(tmpdir)
        if not dist_dir:
            logger.info("Building distributions...")
            build(
                targets,
                rm_destination=True,
                skip_root=False,
                short_names=short_names,
                no_hash=no_hash,
                dist_dir=tmpdir / "dist",
                build_dir=tmpdir / "build",
            )
            dist_dir = tmpdir / "dist"

        out_path.mkdir()
        (out_path / "files").mkdir()

        variables = inspect(targets, short_names=short_names, no_hash=no_hash)
        for scenario in variables["scenarios"]:
            for package in scenario["packages"]:
                # Create a new distributions section
                package["dists"] = []
                for version in package["versions"]:
                    for file in Path(
                        dist_dir
                        / (
                            scenario["name"]
                            if no_hash
                            else f"{scenario['name']}-{scenario['version']}"
                        )
                    ).iterdir():
                        if package["name"].replace("-", "_") + "-" + version[
                            "version"
                        ] not in str(file):
                            continue

                        # Include all the version information to start
                        dist = version.copy()
                        # Then add a sha256
                        dist["sha256"] = sha256_file(file)
                        dist["file"] = file.name
                        package["dists"].append(dist)
                        logger.info(
                            "Found distribution %s",
                            file.name,
                        )
                        file.rename(out_path / "files" / file.name)

        logger.info("Populating template...")
        create_from_template(
            out_path,
            "index",
            variables=variables,
        )

    logger.info("Built static index at ./%s", out_path)
