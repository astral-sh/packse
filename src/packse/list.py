"""
List all scenarios.
"""
import logging
from pathlib import Path
from typing import cast

from packse.error import FileNotFound, InvalidScenario
from packse.scenario import (
    Scenario,
    load_scenarios,
    scenario_hash,
)

logger = logging.getLogger(__name__)


def list(
    targets: list[Path],
    no_hash: bool = False,
    skip_invalid: bool = False,
    no_sources: bool = False,
    short_names: bool = False,
):
    scenarios_by_path: dict[Path, list[Scenario]] = {}

    # Validate and collect all targets first
    for target in sorted(targets):
        if not target.exists():
            raise FileNotFound(target)

        try:
            logger.debug("Loading %s", target)
            scenarios_by_path[target] = load_scenarios(target)
        except Exception as exc:
            if not skip_invalid:
                raise InvalidScenario(target, reason=str(exc)) from exc

    # Then list each one
    for source, scenarios in scenarios_by_path.items():
        prefix = "" if no_sources else " " * 4
        if not no_sources:
            print(to_display_path(source))

        for scenario in scenarios:
            scenario = cast(Scenario, scenario)

            if no_hash:
                name = scenario.name
            else:
                name = scenario_hash(scenario)
                if not short_names:
                    name = f"{scenario.name}-{name}"

            print(prefix + name)


def to_display_path(path: Path | str, relative_to: Path | str | None = None) -> str:
    """
    Convert a path to a displayable path. The absolute path or relative path to the
    current (or given) directory will be returned, whichever is shorter.
    """
    path, relative_to = (
        Path(path).resolve(),
        Path(relative_to or ".").resolve(),
    )

    absolute_path = str(path)

    try:
        relative_path = str(path.relative_to(relative_to))
    except ValueError:
        # Not a child path, use the other one
        return absolute_path

    return relative_path if len(relative_path) < len(absolute_path) else absolute_path
