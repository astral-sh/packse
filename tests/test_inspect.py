import pytest
from packse import __development_base_path__

from .common import snapshot_command


def test_inspect_no_target_finds_all_valid_scenarios(snapshot):
    assert (
        # Just assert it succeeds, we don't want to include all these targets in the snap
        snapshot_command(["inspect"], snapshot_stdout=False, snapshot_stderr=False)
        == snapshot
    )


def test_inspect_target_does_not_exist(snapshot):
    assert snapshot_command(["inspect", "foo"]) == snapshot


@pytest.mark.usefixtures("tmpcwd")
def test_inspect_one_target_does_not_exist(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["inspect", str(target), "foo"]) == snapshot


def test_inspect_invalid_target(snapshot, tmpcwd):
    bad_target = tmpcwd / "test.json"
    bad_target.touch()
    good_target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["inspect", str(bad_target), str(good_target)]) == snapshot


def test_inspect_invalid_target_skip_invalid(snapshot, tmpcwd):
    bad_target = tmpcwd / "test.json"
    bad_target.touch()
    good_target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert (
        snapshot_command(
            ["inspect", str(bad_target), str(good_target), "--skip-invalid"]
        )
        == snapshot
    )


def test_inspect_json(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert snapshot_command(["inspect", str(target)]) == snapshot


def test_inspect_toml(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.toml"
    assert snapshot_command(["inspect", str(target)]) == snapshot


def test_inspect_yaml(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.yaml"
    assert snapshot_command(["inspect", str(target)]) == snapshot
