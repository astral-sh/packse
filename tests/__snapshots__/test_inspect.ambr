# serializer version: 1
# name: test_inspect_invalid_target
  dict({
    'exit_code': 1,
    'stderr': '''
      File at '[PWD]/test.json' is not a valid scenario: Input data was truncated.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_inspect_invalid_target_skip_invalid
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      Skipping file: File at '[PWD]/test.json' is not a valid scenario: Input data was truncated
      {
        "scenarios": [
          {
            "name": "example",
            "packages": [
              {
                "name": "example-a-961b4c22",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-b-961b4c22>1.0.0"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  }
                ]
              },
              {
                "name": "example-b-961b4c22",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "2.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-c-961b4c22"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "3.0.0"
                  }
                ]
              }
            ],
            "root": {
              "requires_python": ">=3.8",
              "requires": [
                {
                  "requirement": "example-a-961b4c22",
                  "name": "example-a-961b4c22",
                  "module_name": "example_a_961b4c22"
                }
              ]
            },
            "expected": {
              "satisfiable": true,
              "packages": [
                {
                  "name": "example-a-961b4c22",
                  "version": "1.0.0",
                  "module_name": "example_a_961b4c22"
                },
                {
                  "name": "example-b-961b4c22",
                  "version": "2.0.0",
                  "module_name": "example_b_961b4c22"
                }
              ],
              "explanation": "The latest valid version of `b` should be installed. `b==3.0.0` is not valid because it requires `c` which does not exist."
            },
            "environment": {
              "python": "3.8",
              "additional_python": []
            },
            "resolver_options": {
              "python": null,
              "prereleases": false,
              "no_build": [],
              "no_binary": [],
              "universal": false
            },
            "template": "package",
            "description": "This is an example scenario, in which the user depends on a single package `a` which requires `b`.",
            "source": "[PROJECT_ROOT]/scenarios/examples/example.json",
            "version": "961b4c22",
            "tree": [
              "\u251c\u2500\u2500 environment",
              "\u2502   \u2514\u2500\u2500 python3.8",
              "\u251c\u2500\u2500 root",
              "\u2502   \u2514\u2500\u2500 requires a",
              "\u2502       \u2514\u2500\u2500 satisfied by a-1.0.0",
              "\u251c\u2500\u2500 a",
              "\u2502   \u2514\u2500\u2500 a-1.0.0",
              "\u2502       \u2514\u2500\u2500 requires b>1.0.0",
              "\u2502           \u251c\u2500\u2500 satisfied by b-2.0.0",
              "\u2502           \u2514\u2500\u2500 satisfied by b-3.0.0",
              "\u2514\u2500\u2500 b",
              "    \u251c\u2500\u2500 b-1.0.0",
              "    \u251c\u2500\u2500 b-2.0.0",
              "    \u2514\u2500\u2500 b-3.0.0",
              "        \u2514\u2500\u2500 requires c",
              "            \u2514\u2500\u2500 unsatisfied: no versions for package"
            ],
            "module_name": "example"
          }
        ]
      }
  
    ''',
  })
# ---
# name: test_inspect_json
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      {
        "scenarios": [
          {
            "name": "example",
            "packages": [
              {
                "name": "example-a-961b4c22",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-b-961b4c22>1.0.0"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  }
                ]
              },
              {
                "name": "example-b-961b4c22",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "2.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-c-961b4c22"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "3.0.0"
                  }
                ]
              }
            ],
            "root": {
              "requires_python": ">=3.8",
              "requires": [
                {
                  "requirement": "example-a-961b4c22",
                  "name": "example-a-961b4c22",
                  "module_name": "example_a_961b4c22"
                }
              ]
            },
            "expected": {
              "satisfiable": true,
              "packages": [
                {
                  "name": "example-a-961b4c22",
                  "version": "1.0.0",
                  "module_name": "example_a_961b4c22"
                },
                {
                  "name": "example-b-961b4c22",
                  "version": "2.0.0",
                  "module_name": "example_b_961b4c22"
                }
              ],
              "explanation": "The latest valid version of `b` should be installed. `b==3.0.0` is not valid because it requires `c` which does not exist."
            },
            "environment": {
              "python": "3.8",
              "additional_python": []
            },
            "resolver_options": {
              "python": null,
              "prereleases": false,
              "no_build": [],
              "no_binary": [],
              "universal": false
            },
            "template": "package",
            "description": "This is an example scenario, in which the user depends on a single package `a` which requires `b`.",
            "source": "[PWD]/scenarios/examples/example.json",
            "version": "961b4c22",
            "tree": [
              "\u251c\u2500\u2500 environment",
              "\u2502   \u2514\u2500\u2500 python3.8",
              "\u251c\u2500\u2500 root",
              "\u2502   \u2514\u2500\u2500 requires a",
              "\u2502       \u2514\u2500\u2500 satisfied by a-1.0.0",
              "\u251c\u2500\u2500 a",
              "\u2502   \u2514\u2500\u2500 a-1.0.0",
              "\u2502       \u2514\u2500\u2500 requires b>1.0.0",
              "\u2502           \u251c\u2500\u2500 satisfied by b-2.0.0",
              "\u2502           \u2514\u2500\u2500 satisfied by b-3.0.0",
              "\u2514\u2500\u2500 b",
              "    \u251c\u2500\u2500 b-1.0.0",
              "    \u251c\u2500\u2500 b-2.0.0",
              "    \u2514\u2500\u2500 b-3.0.0",
              "        \u2514\u2500\u2500 requires c",
              "            \u2514\u2500\u2500 unsatisfied: no versions for package"
            ],
            "module_name": "example"
          }
        ]
      }
  
    ''',
  })
# ---
# name: test_inspect_no_target_finds_all_valid_scenarios
  dict({
    'exit_code': 0,
    'stderr': '<not included>',
    'stdout': '<not included>',
  })
# ---
# name: test_inspect_one_target_does_not_exist
  dict({
    'exit_code': 1,
    'stderr': '''
      File 'foo' not found.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_inspect_target_does_not_exist
  dict({
    'exit_code': 1,
    'stderr': '''
      File 'foo' not found.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_inspect_toml
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      {
        "scenarios": [
          {
            "name": "example-toml",
            "packages": [
              {
                "name": "example-toml-a-4a4ede4d",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-toml-b-4a4ede4d>1.0.0"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  }
                ]
              },
              {
                "name": "example-toml-b-4a4ede4d",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "2.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-toml-c-4a4ede4d"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "3.0.0"
                  }
                ]
              }
            ],
            "root": {
              "requires_python": ">=3.8",
              "requires": [
                {
                  "requirement": "example-toml-a-4a4ede4d",
                  "name": "example-toml-a-4a4ede4d",
                  "module_name": "example_toml_a_4a4ede4d"
                }
              ]
            },
            "expected": {
              "satisfiable": true,
              "packages": [
                {
                  "name": "example-toml-a-4a4ede4d",
                  "version": "1.0.0",
                  "module_name": "example_toml_a_4a4ede4d"
                },
                {
                  "name": "example-toml-b-4a4ede4d",
                  "version": "2.0.0",
                  "module_name": "example_toml_b_4a4ede4d"
                }
              ],
              "explanation": "The latest valid version of `b` should be installed. `b==3.0.0` is not valid because it requires `c` which does not exist."
            },
            "environment": {
              "python": "3.8",
              "additional_python": []
            },
            "resolver_options": {
              "python": null,
              "prereleases": false,
              "no_build": [],
              "no_binary": [],
              "universal": false
            },
            "template": "package",
            "description": "This is an example scenario written in TOML, in which the user depends on a single package `a` which requires `b`.",
            "source": "[PWD]/scenarios/examples/example.toml",
            "version": "4a4ede4d",
            "tree": [
              "\u251c\u2500\u2500 environment",
              "\u2502   \u2514\u2500\u2500 python3.8",
              "\u251c\u2500\u2500 root",
              "\u2502   \u2514\u2500\u2500 requires a",
              "\u2502       \u2514\u2500\u2500 satisfied by a-1.0.0",
              "\u251c\u2500\u2500 a",
              "\u2502   \u2514\u2500\u2500 a-1.0.0",
              "\u2502       \u2514\u2500\u2500 requires b>1.0.0",
              "\u2502           \u251c\u2500\u2500 satisfied by b-2.0.0",
              "\u2502           \u2514\u2500\u2500 satisfied by b-3.0.0",
              "\u2514\u2500\u2500 b",
              "    \u251c\u2500\u2500 b-1.0.0",
              "    \u251c\u2500\u2500 b-2.0.0",
              "    \u2514\u2500\u2500 b-3.0.0",
              "        \u2514\u2500\u2500 requires c",
              "            \u2514\u2500\u2500 unsatisfied: no versions for package"
            ],
            "module_name": "example_toml"
          }
        ]
      }
  
    ''',
  })
# ---
# name: test_inspect_yaml
  dict({
    'exit_code': 0,
    'stderr': '',
    'stdout': '''
      {
        "scenarios": [
          {
            "name": "example-yaml",
            "packages": [
              {
                "name": "example-yaml-a-4692130b",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-yaml-b-4692130b>1.0.0"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  }
                ]
              },
              {
                "name": "example-yaml-b-4692130b",
                "versions": [
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "1.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "2.0.0"
                  },
                  {
                    "requires_python": ">=3.8",
                    "requires": [
                      "example-yaml-c-4692130b"
                    ],
                    "extras": [],
                    "sdist": true,
                    "wheel": true,
                    "yanked": false,
                    "wheel_tags": [],
                    "description": "",
                    "version": "3.0.0"
                  }
                ]
              }
            ],
            "root": {
              "requires_python": ">=3.8",
              "requires": [
                {
                  "requirement": "example-yaml-a-4692130b",
                  "name": "example-yaml-a-4692130b",
                  "module_name": "example_yaml_a_4692130b"
                }
              ]
            },
            "expected": {
              "satisfiable": true,
              "packages": [
                {
                  "name": "example-yaml-a-4692130b",
                  "version": "1.0.0",
                  "module_name": "example_yaml_a_4692130b"
                },
                {
                  "name": "example-yaml-b-4692130b",
                  "version": "2.0.0",
                  "module_name": "example_yaml_b_4692130b"
                }
              ],
              "explanation": "The latest valid version of `b` should be installed. `b==3.0.0` is not valid because it requires `c` which does not exist."
            },
            "environment": {
              "python": "3.8",
              "additional_python": []
            },
            "resolver_options": {
              "python": null,
              "prereleases": false,
              "no_build": [],
              "no_binary": [],
              "universal": false
            },
            "template": "package",
            "description": "This is an example scenario written in YAML, in which the user depends on a single package `a` which requires `b`.",
            "source": "[PWD]/scenarios/examples/example.yaml",
            "version": "4692130b",
            "tree": [
              "\u251c\u2500\u2500 environment",
              "\u2502   \u2514\u2500\u2500 python3.8",
              "\u251c\u2500\u2500 root",
              "\u2502   \u2514\u2500\u2500 requires a",
              "\u2502       \u2514\u2500\u2500 satisfied by a-1.0.0",
              "\u251c\u2500\u2500 a",
              "\u2502   \u2514\u2500\u2500 a-1.0.0",
              "\u2502       \u2514\u2500\u2500 requires b>1.0.0",
              "\u2502           \u251c\u2500\u2500 satisfied by b-2.0.0",
              "\u2502           \u2514\u2500\u2500 satisfied by b-3.0.0",
              "\u2514\u2500\u2500 b",
              "    \u251c\u2500\u2500 b-1.0.0",
              "    \u251c\u2500\u2500 b-2.0.0",
              "    \u2514\u2500\u2500 b-3.0.0",
              "        \u2514\u2500\u2500 requires c",
              "            \u2514\u2500\u2500 unsatisfied: no versions for package"
            ],
            "module_name": "example_yaml"
          }
        ]
      }
  
    ''',
  })
# ---
